File: lab7/Dictionary.java
/**
 * File: Dictionary.java
 * Name: Sophia Tacderas
 * Due: 12/2/16, 10 pm
 * Class: CMPS 12M
 * Assignment: lab 7
 * Purpose: Implements the Dictionary interface with a binary search tree.
 * Borrows/modifies code from: Dictionary.c, pa3
 */

public class Dictionary implements DictionaryInterface {

    // private types and functions ------------------------------------------------

    // private inner Node class
    private class Node {
        // key is distinct/unique
        String key;
        // value may repeat
        String value;
        // pointer to left and right Nodes
        Node leftNode;
        Node rightNode;

        // constructor for private Node type
        Node(String x, String y) {
            key = x;
            value = y;
            leftNode = null;
            rightNode = null;
        }
    }

    // Private fields for the Dictionary class
    private Node root; // reference to root Node in Dictionary BST
    private int numPairs; // number of pairs in this Dictionary

    // findKey()
    // returns the Node containing key k in the subtree rooted at R, or returns
    // null if no such Node exists
    private Node findKey(Node R, String k) {
        if (R == null || k.compareTo(R.key) == 0) {
            return R;
        }
        if (k.compareTo(R.key) < 0) {
            return findKey(R.leftNode, k);
        } else { // k.compareTo(R.key) > 0
            return findKey(R.rightNode, k);
        }
    }

    // findParent()
    // returns the parent of N in the subtree rooted at R, or returns null
    // if N is equal to R. (pre: R != null)
    private Node findParent(Node N, Node R) {
        Node P = null;
        if (N != R) {
            P = R;
            while ((P.leftNode != N) && (P.rightNode != N)) {
                if ((P.key).compareTo(R.key) < 0) {
                    P = P.leftNode;
                } else { // (P.key).compareTo(R.key) > 0
                    P = P.rightNode;
                }
            }
        }
        return P;
    }

    // findLeftmost()
    // returns the leftmost Node in the subtree rooted at R, or null if R is null.
    private Node findLeftmost(Node R) {
        Node L = R;
        if (L != null) {
            for (; L.leftNode != null; L = L.leftNode) ;
        }
        return L;
    }

    // printInOrder()
    // prints the (key, value) pairs belonging to the subtree rooted at R in order
    // of increasing keys to file pointed to by out.
    private void printInOrder(Node R) {
        if (R != null) {
            printInOrder(R.leftNode);
            System.out.println(R.key + " " + R.value);
            printInOrder(R.rightNode);
        }
    }

    // deleteAll()
    // deletes all Nodes in the subtree rooted at N.
    private void deleteAll(Node N) {
        if (N != null) {
            deleteAll(N.leftNode);
            deleteAll(N.rightNode);
        }
    }

    // public functions -----------------------------------------------------------

    // Constructor for the Dictionary class
    public Dictionary() {
        root = null;
        numPairs = 0;
    }

    // isEmpty()
    // pre: none
    // returns true if this Dictionary is empty, false otherwise
    public boolean isEmpty() {
        if (numPairs == 0) {
            return true;
        } else {
            return false;
        }
    }

    // size()
    // pre: none
    // returns the number of entries in this Dictionary
    public int size() {
        return numPairs;
    }

    // lookup()
    // pre: none
    // returns value associated key, or null reference if no such key exists
    public String lookup(String key) {
        Node N;
        N = findKey(root, key);
        if (N != null) {
            return N.value;
        } else {
            return null;
        }
    }

    // insert()
    // inserts new (key,value) pair into this Dictionary
    // pre: lookup(key)==null
    public void insert(String key, String value) throws DuplicateKeyException {
        Node N, A, B;
        if (lookup(key) != null) {
            throw new DuplicateKeyException(
                    "Dictionary Error: cannot insert() duplicate key: \"" + key + "\"");
        }
        N = new Node(key, value);
        B = null;
        A = root;
        while (A != null) {
            B = A;
            if (key.compareTo(A.key) < 0) {
                A = A.leftNode;
            } else { // key.compareTo(A.key) > 0
                A = A.rightNode;
            }
        }
        if (B == null) {
            root = N;
        } else if (key.compareTo(B.key) < 0) {
            B.leftNode = N;
        } else { // key.compareTo(B.key) > 0
            B.rightNode = N;
        }
        numPairs++;
    }

    // delete()
    // deletes pair with the given key
    // pre: lookup(key)!=null
    public void delete(String key) throws KeyNotFoundException {
        Node N, P, S;
        if (lookup(key) == null) {
            throw new KeyNotFoundException(
                    "Dictionary Error: cannot delete() non-existent key: \"" + key + "\"");
        } else {
            N = findKey(root, key);
            if (N.leftNode == null && N.rightNode == null) { // case 1 (no children)
                if (N == root) {
                    root = null;
                } else { // N is not the root
                    P = findParent(N, root);
                    if (P.rightNode == N) {
                        P.rightNode = null;
                    } else { // P.leftNode == N
                        P.leftNode = null;
                    }
                }
            } else if (N.rightNode == null) { // case 2 (left but no right child)
                if (N == root) {
                    root = N.leftNode;
                } else { // N is not the root
                    P = findParent(N, root);
                    if (P.rightNode == N) {
                        P.rightNode = N.leftNode;
                    } else { // P.leftNode == N
                        P.leftNode = N.leftNode;
                    }
                }
            } else if (N.leftNode == null) { // case 2 (right but no left child)
                if (N == root) {
                    root = N.rightNode;
                } else { // N is not the root
                    P = findParent(N, root);
                    if (P.rightNode == N) {
                        P.rightNode = N.rightNode;
                    } else { // P.leftNode == N
                        P.leftNode = N.rightNode;
                    }
                }
            } else { // case 3 (two children: N.leftNode!=null && N.rightNode!=null)
                S = findLeftmost(N.rightNode);
                P = findParent(S, N);
                N.key = S.key;
                N.value = S.value;
                if (P.rightNode == S) {
                    P.rightNode = S.rightNode;
                } else { // P.leftNode == S
                    P.leftNode = S.rightNode;
                }
            }
            numPairs--;
        }
    }

    // makeEmpty()
    // pre: none
    public void makeEmpty() {
        deleteAll(root);
        root = null;
        numPairs = 0;
    }

    // toString()
    // returns a String representation of this Dictionary
    // overrides Object's toString() method
    // pre: none
    public String toString() {
        printInOrder(root);
        return "";
    }

}

File: pa5/Dictionary.c
/**
 * File: Dictionary.c
 * Name: Sophia Tacderas
 * ID: 1465379
 * Due: 12/2/16, 10 pm
 * Class: CMPS 12B
 * Assignment: pa5
 * Purpose: Implements Dictionary.h with using a hash table.
 * Borrows/modifies code from: Dictionary.c, lab5
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "Dictionary.h"

const int tableSize = 101; // constant table size

// private types --------------------------------------------------------------

// NodeObj
typedef struct NodeObj{
    char* key;
    char* value;
    struct NodeObj* next;
} NodeObj;

// Node
typedef NodeObj* Node;

// newNode()
// constructor of the Node type
Node newNode(char* k, char* v) {
    Node N = malloc(sizeof(NodeObj));
    assert(N!=NULL);
    N->key = k;
    N->value = v;
    N->next = NULL;
    return(N);
}

// freeNode()
// destructor for the Node type
void freeNode(Node* pN){
    if( pN!=NULL && *pN!=NULL ){
        free(*pN);
        *pN = NULL;
    }
}

// DictionaryObj
typedef struct DictionaryObj{
    /*
     * Several methods for resolving collisions will be discussed in class. The method that we will use in this
     * project is called chaining, and is perhaps the simplest collision resolution technique. In chaining, we put
     * all the pairs that hash to the same slot into a linked list.
     *
     * Thus the hash table T[0…(m −1)] is an array of
     * linked lists. More precisely, T[ j] is a pointer to the head of a linked list storing all pairs that hash to slot
     * j. If there are no such elements, T[ j] will be NULL.
     */
    Node* hashTable;
    int numItems;
} DictionaryObj;


// public functions -----------------------------------------------------------

/*
 * Functions rotate_left() and pre_hash() turn a string into an unsigned int, and function
 * hash() converts that number into an int in the range 0 to tableSize −1.
 */
// rotate_left()
// rotate the bits in an unsigned int
unsigned int rotate_left(unsigned int value, int shift) {
    int sizeInBits = 8*sizeof(unsigned int);
    shift = shift & (sizeInBits - 1);
    if ( shift == 0 )
        return value;
    return (value << shift) | (value >> (sizeInBits - shift));
}

// pre_hash()
// turn a string into an unsigned int
unsigned int pre_hash(char* input) {
    unsigned int result = 0xBAE86554;
    while (*input) {
        result ^= *input++;
        result = rotate_left(result, 5);
    }
    return result;
}

// hash()
// turns a string into an int in the range 0 to tableSize-1
int hash(char* key){
    return pre_hash(key)%tableSize;
}

// newDictionary()
// constructor for the Dictionary type
Dictionary newDictionary(void){
    Dictionary myDict = malloc(sizeof(DictionaryObj));
    assert(myDict!=NULL);
    myDict->hashTable = calloc(tableSize, sizeof(Node));
    assert(myDict->hashTable != NULL);
    myDict->numItems = 0;
    return myDict;
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD){
    if( pD!=NULL && *pD!=NULL ){
        if( !isEmpty(*pD) ){
            makeEmpty(*pD);
        }
        free((*pD)->hashTable); // free hash table
        free(*pD); // free DictionaryObj
        *pD = NULL;
    }
}

// findKey()
// Returns a reference to the Node containing its argument key from the hash table.
// Otherwise, returns NULL if that Node does not exist.
Node findKey(Dictionary D, char* key) {
    int intKey = hash(key); // get int representation of argument key
    Node N = D->hashTable[intKey]; // find Node for equivalent key in hash table
    /*
    * To lookup a given key k, we do a linear search of the list T[h(k)], and return the corresponding value if
    * found, or NULL if not found.
    */
    if(N != NULL) {
        int ctr = 0;
        for (ctr = 1; ctr <= D->numItems; ctr++) {
            if (strcmp(N->key, key)==0){
                return N;
            } else {
                N = N->next;
            }
        }
    }
    return NULL;
}

// isEmpty()
// returns 1 (true) if D is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D){
    if( D==NULL ){
        fprintf(stderr,
                "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
    return(D->numItems==0);
}

// size()
// returns the number of (key, value) pairs in D
// pre: none
int size(Dictionary D) {
    return D->numItems;
}

// lookup()
// returns the value v such that (k, v) is in D, or returns NULL if no
// such value v exists.
// pre: none
char* lookup(Dictionary D, char* k){
    // use findKey() to find the Node
    Node N = findKey(D, k);
    if (N != NULL) {
        return N->value;
    } else {
        return NULL;
    }
}

// insert()
// inserts new (key,value) pair into D
// pre: lookup(D, k)==NULL
void insert(Dictionary D, char* k, char* v){
    if(lookup(D, k) == NULL) { // arguments are not duplicate
        // hash argument key
        int intKey = hash(k);

        // find Node for equivalent key in hash table
        Node N = D->hashTable[intKey];

        if (N == NULL) {
            // table is empty so this is first item
            /*
            * To insert a pair (k, v) into the Dictionary,
            * we create a new Node storing this pair, then insert that Node at the head of the linked list T[h(k)].
            */
            Node insertNode = newNode(k, v);
            insertNode->next = D->hashTable[intKey];
            D->hashTable[intKey] = insertNode;
        } else {
            // insert item at the end of list
            int ctr = 0;
            for (ctr = 1; ctr < D->numItems; ctr++) {
                N = N->next;
            }
            N->next = newNode(k, v);
        }
        D->numItems++;

    } else {
        fprintf(stderr, "Calling cannot insert duplicate keys: %s\n", k);
        exit(EXIT_FAILURE);
    }
}

// delete()
// deletes pair with the key k
// pre: lookup(D, k)!=NULL
void delete(Dictionary D, char* k){
    if( lookup(D, k)==NULL ){
        fprintf(stderr, "Dictionary Error: delete() cannot delete non-existent key.\n");
        exit(EXIT_FAILURE);
    }

    // hash argument key
    int intKey = hash(k);

    // find Node for equivalent key in hash table
    Node N = D->hashTable[intKey];

    /*
    * To delete the pair with key k, simply splice the corresponding Node out of the list headed by T[h(k)].
    */
    if (strcmp(N->key, k) == 0) {
        D->hashTable[intKey] = N->next;
        N->next = NULL;
        freeNode(&N);
    } else {
        Node prev = N;
        Node temp = N->next;

        while (temp->key != k) {
            temp = temp->next;
            prev = prev->next;
        }

        prev->next = temp->next;
        temp->next = NULL;

        freeNode(&temp);
    }
    D->numItems--;
}

// makeEmpty()
// re-sets D to the empty state.
// pre: none
void makeEmpty(Dictionary D){
    if( D==NULL ){
        fprintf(stderr,
                "Dictionary Error: calling makeEmpty() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
    if( D->numItems==0 ){
        fprintf(stderr, "Dictionary Error: calling makeEmpty() on empty Dictionary\n");
        exit(EXIT_FAILURE);
    }
    int ctr = 0;
    for (ctr = 0; ctr < tableSize; ctr++) {
        Node N = D->hashTable[ctr];
        if (N != NULL) {
            Node temp = N->next;
            D->hashTable[ctr] = temp;
            freeNode(&N);
        }
    }
    D->numItems = 0;
}

// printDictionary()
// pre: none
// prints a text representation of D to the file pointed to by out
void printDictionary(FILE* out, Dictionary D){
    /*
     * Your printDictionary() function in this project should
     * just print out pairs in the order that they appear in the table, i.e. print out list T[0] in order
     * (i.e. head to tail), then list T[1], …, then end by printing list T[m −1].
     */
    if( D==NULL ){
        fprintf(stderr,
                "Dictionary Error: calling printDictionary() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
    int ctr = 0;
    for (ctr = 0; ctr < tableSize; ctr++) {
        Node N = D->hashTable[ctr];
        if (N != NULL) {
            fprintf(out, "%s %s\n", N->key, N->value);
        }
    }
}
